pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+// can be a partial one.
+func (m *Mesh) UpdateSubsets(subsetEvents []SubsetEvent) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, ssEvent := range subsetEvents {
+		subset := ssEvent.Subset
+		if ssEvent.Event == EventDelete || ssEvent.Event == EventUpdate {
+			ssName := subset.Name
+			oldEpKeySet := m.subsetEndpoints[ssName]
+			for epKey := range oldEpKeySet {
+				m.reverseEpSubsets.deleteSubset(epKey, ssName)
+			}
+			delete(m.subsetEndpoints, ssName)
+		}
+		if ssEvent.Event == EventAdd || ssEvent.Event == EventUpdate {
+			ssName := subset.Name
+			epUIDSet := m.reverseAttrMap.getKeysMatching(subset.Labels)
+			m.subsetEndpoints.addEndpointUIDs(ssName, epUIDSet)
+			for endpointUID := range epUIDSet {
+				m.reverseEpSubsets.addSubset(endpointUID, ssName)
+			}
+		}
+	}
+}
+
+// NewEndpoint is a boiler plate function intended for platform Controllers to create a new Endpoint.
+// This method ensures all the necessary data required for creating subsets are correctly setup. It
+// also performs sorting of arrays etc, to allow stable results for reflect.DeepEquals() for quick
+// comparisons.
+func NewEndpoint(endpointUID, service string, aliases []string, address string, port uint32, protocol Protocol, labels Labels, user string) (*Endpoint, error) {
+	var errs error
+	ipAddr := net.ParseIP(address)
+	if ipAddr == nil {
+		errs = multierror.Append(errs, errors.New(fmt.Sprintf("invalid IP address '%s'", address)))
+	}
+	svcParts := svcRegexp.FindStringSubmatch(service)
+	if svcParts == nil || len(svcParts) != 3 {
+		errs = multierror.Append(
+			errors.New(fmt.Sprintf(
+				"invalid service name format, expecting form <name>.<namespace>.svc.cluster.local found'%s'",
+				service)))
+	}
+	for _, alias := range aliases {
+		if !domainRegexp.MatchString(alias) {
+			errs = multierror.Append(errors.New(
+				fmt.Sprintf("invalid service alias format for service '%s', expecting format '%s' found'%s'",
+					service, domainRegexp.String(), alias)))
+		}
+	}
+	destLabels := make(Labels, len(labels)+len(destinationAttrSet))
+	for labelName, labelValue := range labels {
+		if isDestinationAttribute(labelName) {
+			errs = multierror.Append(errors.New(
+				fmt.Sprintf("prohibited use of Istio destination label '%s' for endpoint label", labelName)))
+		}
+		destLabels[labelName] = labelValue
+	}
+	if errs != nil {
+		return nil, multierror.Prefix(errs, "Mesh endpoint creation errors")
+	}
+	socketProtocol := xdsapi.SocketAddress_TCP
+	switch protocol {
+	case ProtocolTCP, ProtocolHTTP, ProtocolHTTP2, ProtocolHTTPS, ProtocolRedis, ProtocolMongo, ProtocolGRPC:
+		socketProtocol = xdsapi.SocketAddress_TCP
+	case ProtocolUDP:
+		socketProtocol = xdsapi.SocketAddress_UDP
+	}
+	ep := Endpoint{
+		Endpoint: &xdsapi.Endpoint{
+			Address: &xdsapi.Address{
+				Address: &xdsapi.Address_SocketAddress{
+					&xdsapi.SocketAddress{
+						Address:    address,
+						Ipv4Compat: ipAddr.To4() != nil,
+						Protocol:   socketProtocol,
```

Comment:
cc @rshriram : 
Is this implying exchange of cluster-scoped IPv4 addresses between clusters?

Reply:
You mean the "Address" ?Yes, thats a good catch. Doesn't that imply a flatnetwork by default?
I do not see how the subsets can be materialized without breaking the abstraction we have today (route rules are processed by proxy/envoy, while platform only computes the default clusters. Platform has no idea about routing rules.

I also dont see how we would go about the zero vpn case where each cluster/pilot is only going to expose its ingress as the endpoint for all clusters, because the platform code has no idea whether the endpoints its giving out is going to be consumed by local pilot or a remote pilot.

Syncing clusters across two meshes can be done in a different way (thinking out aloud). If you imagine that the code under dataplane/ is exposing a variant of CDS/EDS api (e.g., /v2/remote/envoy.api.v2.clusterblahblah, then a remote pilot can query the local pilot and get the list of clusters. When pilot gives out the clusters, it could empty out the endpoints and set them all to just ingress or use the actual endpoints depending on the flat network presence. The remote pilot will simply need to merge the locally computed cluster with the remote clusters and serve them up to the envoys. None of this has to be aware of route rules/services or any such thing. Essentially, rather than putting this code in model and trying to have two "views", its easier to stick this directly in the data plane as an API for the remote clusters to query. This won't break the clean separation of service registry and routing data we have. Otherwise, the code is going to be messy (proxy/envoy/* will start writing into the model registry or we end up moving some route computation out of proxy/envoy/* into platform/, which is not sitting well with me due to the multitude of cases involved.)

Thoughts? (@costinm / @kyessenov )

Reply:
See comments above in the file. I've avoided using the term flat or non-flat. Network reachable is the key here. Any endpoint exposed by a remote-pilot-A to remote-pilot-B should be reachable from remote-pilot-B.

Reply:
On IPv4 implication ... the Address can be IPv4 OR IPv6. For label matching we always normalize the IP address. So multiple address representations supplied by various platforms will be queried and matched correctly. 

This particular flag is for whether the IPv6 is IPv4 compatible or not. I don't need to populate this flag.

Reply:
Yes, I am well aware of the comment. Despite that comment, what you are doing here is setting local endpoint IP address (ip that is not visible outside in non flat networks), and exporting this around everywhere. And my question was not about reachability. It was abotu the way this code was designed.

Reply:
"local endpoint IP address (ip that is not visible outside in non flat networks)" this __not__ what this current PR does. That part of the code is yet to be written. But upcoming PRs will show that Pilots will only expose reachable addresses. i.e. it will expose the gateway's IP rather than the  endpoints that are backed by that gateway.

Reply:
Yeah but these are coming from the platforms that have no idea who is going to consume these ips. In mesh or cross mesh.

Reply:
I don't see a real issue here. If the platform contains both reachable and unreachable endpoints, this is something we can address in the following PR which implements exporting endpoints to remote pilots. The export interface is can ascertain Pilot's ingress / egress knowledge and perform the necessary translation before exporting the endpoints.